File: 2015/python/astar.py
from queue import PriorityQueue
import functools

# https://bugs.python.org/issue31145
@functools.total_ordering
class Prio:
    def __init__(self, f, node):
        self.f = f
        self.node = node

    def __eq__(self, other):
        return self.f == other.f

    def __lt__(self, other):
        return self.f < other.f


def reconstruct_path(came_from, current):
    path = [current]
    while current in came_from:
        current = came_from[current]
        path.append(current)
    path.reverse()
    return path


# https://en.wikipedia.org/wiki/A*_search_algorithm
def astar(start, adjacent, h, is_target) -> list:
    open_queue = PriorityQueue()
    open_queue.put(Prio(0, start))
    open_set = set([start])
    came_from = dict()
    g_score = {start: 0}

    while not open_queue.empty():
        current = open_queue.get()
        open_set.remove(current.node)

        if is_target(current.node):
            return reconstruct_path(came_from, current.node)

        for time, child in adjacent(current.node):
            g = g_score[current.node] + time
            if child not in g_score or g < g_score[child]:
                came_from[child] = current.node
                g_score[child] = g
                f = g + h(child)
                if child not in open_set:
                    open_set.add(child)
                    open_queue.put(Prio(f, child))

File: 2015/rust/lib/src/astar.rs
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap, HashSet};
use std::hash::Hash;

#[derive(Copy, Clone, Eq, PartialEq)]
struct NodeWithCost<N: Eq> {
    cost: usize,
    node: N,
}

impl<N: Eq> Ord for NodeWithCost<N> {
    fn cmp(&self, other: &Self) -> Ordering {
        other.cost.cmp(&self.cost)
    }
}

impl<N: Eq> PartialOrd for NodeWithCost<N> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn reconstruct_path<'a, N: Eq + Hash + Copy>(
    came_from: HashMap<N, N>,
    node: &'a N,
) -> Vec<N> {
    let mut path = Vec::new();
    while let Some(node) = came_from.get(&node) {
        path.push(*node);
    }
    path.reverse();
    return path;
}

// https://en.wikipedia.org/wiki/A*_search_algorithm
pub fn astar<'a, N: Eq + Hash + Copy>(
    start: &'a N,
    adjacent: impl Fn(&N) -> Vec<(usize, N)>,
    h: impl Fn(&N) -> usize,
    is_target: impl Fn(&N) -> bool,
) -> Vec<N> {
    let mut open_queue = BinaryHeap::new();
    let mut open_set = HashSet::from([*start]);
    let mut g_score = HashMap::from([(*start, 0)]);
    let mut came_from = HashMap::new();
    open_queue.push(NodeWithCost {
        cost: 0,
        node: *start,
    });

    while let Some(NodeWithCost { cost: _, node }) = open_queue.pop() {
        if is_target(&node) {
            return reconstruct_path(came_from, &node);
        }

        open_set.remove(&node);

        for (time, child) in adjacent(&node) {
            let g = g_score[&node] + time;
            if !g_score.contains_key(&child) || g < *g_score.get(&child).unwrap() {
                g_score.insert(child, g);
                let f = g + h(&child);
                if !open_set.contains(&child) {
                    open_set.insert(child);
                    open_queue.push(NodeWithCost {
                        cost: f,
                        node: child,
                    });
                }
                came_from.insert(child, node);
            }
        }
    }

    unreachable!("The loop should always return");
}
